src/api/XRSession.js

```diff
@@ -259,59 +354,30 @@ export default class XRSession extends EventTarget {
       return;
     }
 
-    // If the session is suspended and we have a previously saved
-    // suspendedCallback, abort this call
-    if (this[PRIVATE].suspended && this[PRIVATE].suspendedCallback) {
-      return;
-    }
-
-    // Otherwise, if the session is suspended but has not yet creating
-    // the suspended callback, do so. It will resume once it is no
-    // longer suspended.
-    if (this[PRIVATE].suspended && !this[PRIVATE].suspendedCallback) {
-      this[PRIVATE].suspendedCallback = callback;
+    // Add callback to the queue and return its handle
+    const handle = ++this[PRIVATE].frameHandle;
+    const callbacks = this[PRIVATE].frameCallbacks;
+    if (callbacks.findIndex(d => d.callback === callback) === -1) {
+      this[PRIVATE].frameCallbacks.push({
+        handle,
+        callback,
+        cancelled: false
+      });
     }
-
-    // TODO: Should pending render state be applied before or after onFrameStart?
-    return this[PRIVATE].device.requestAnimationFrame(() => {
-      if (this[PRIVATE].pendingRenderState !== null) {
-        // Apply pending render state.
-        this[PRIVATE].activeRenderState = new XRRenderState(this[PRIVATE].pendingRenderState);
-        this[PRIVATE].pendingRenderState = null;
-
-        // TODO: set compositionDisabled
-
-        // Report to the device since it'll need to handle the layer for rendering.
-        if (this[PRIVATE].activeRenderState.baseLayer) {
-          this[PRIVATE].device.onBaseLayerSet(
-            this[PRIVATE].id,
-            this[PRIVATE].activeRenderState.baseLayer);
-        }
-
-        if (this[PRIVATE].activeRenderState.inlineVerticalFieldOfView) {
-          this[PRIVATE].device.onInlineVerticalFieldOfViewSet(
-            this[PRIVATE].id,
-            this[PRIVATE].activeRenderState.inlineVerticalFieldOfView);
-        }
-      }
-      this[PRIVATE].device.onFrameStart(this[PRIVATE].id);
-      // inputSources can be populated in .onFrameStart()
-      // so check the change and fire inputsourceschange event if needed
-      this._checkInputSourcesChange();
-      callback(now(), this[PRIVATE].frame);
-      this[PRIVATE].device.onFrameEnd(this[PRIVATE].id);
-    });
+    return handle;
   }
 
   /**
    * @param {number} handle
    */
   cancelAnimationFrame(handle) {
-    if (this[PRIVATE].ended) {
-      return;
+    // Remove the callback with that handle from the queue
+    const callbacks = this[PRIVATE].frameCallbacks;
+    const index = callbacks.findIndex(d => d.handle === handle);
+    if (index > -1) {
+      callbacks[index].cancelled = true;
+      callbacks.splice(index, 1)
```

Comment:
I'm not gonna block you on this, because it's a SUPER annoying nitpick. but this should also search through the list of currently executing callbacks in `onDeviceFrame()` if there is one. This is because if I have two callbacks queued up, A and B (in that order), and cancel B while A is running it should still respect that cancellation.

In reality that's not gonna happen too often in real-world apps, but it is necessary for correctness. Sorry. 🙄 (And again, I won't block this PR on that. I can fix it in a follow up if needed.)

Reply:
Not an annoying nitpick at all, and easy enough to do.

Does this warrant clarification in the spec? It says to `Find the entry in session’s list of animation frame callbacks that is associated with the value handle`, but *session’s list of animation frame callbacks* was explicitly set to the empty list in the XR animation frame algorithm prior to callback executions.

Reply:
That's a good point. I believe it is worth clarifying.

Reply:
I opened https://github.com/immersive-web/webxr/issues/923 as a followup.